base-files: upgrade: use zcat command provided by busybox

fanck0605 opened the pull request:
If gzip is installed, the [`switch_to_ramfs()`](https://github.com/openwrt/openwrt/blob/3d12b47985fc1983849925d2dc23430f55210c80/package/base-files/files/lib/upgrade/stage2#L35) call will not copy the gzip executable to ramfs, however `/bin/zcat` will call it instead of the busybox-supplied zcat. This will cause sysupgrade to fail.

Signed-off-by: Chuck <fanck0605@qq.com>

fix openwrt/packages#14925

tested on FriendlyElec NanoPi R2S

fanck0605 commented:
This patch only fixes some of the problems, and there may be some similar problems with sysupgrade that we haven't found yet.

I think that during sysupgrade we should only call the methods provided by busybox and not trust any executable provided by the system directory

adschm commented on package/base-files/files/lib/upgrade/common.sh:
```diff
@@ -101,7 +101,7 @@ get_image() { # <source> [ <command> ]
 	fi
 
-	$cmd <"$from"
+	busybox $cmd <"$from"
```
If you only want to alter a specific command, this should be changed in the switch case above instead.

fanck0605 replied:
The `cat` command is originally provided by busybox, and busybox also provides the `bzcat` command. `bzcat` may also cause the same problems because of the installation of `bzip`, so adding the busybox prefix here can fix these problems together.

adschm commented:
This lacks a commit message and Signed-off-by should contain the full name.

fanck0605 commented:
@adschm modified

zhanhb reviewed (changes requested):
IMO, work with busybox zcat is enough. bzcat is a symlink to `/usr/bin/bzip2`, which can work as expect.

```patch
diff --git a/package/base-files/files/lib/upgrade/common.sh b/package/base-files/files/lib/upgrade/common.sh
index 968893e226..31ffc3cf03 100644
--- a/package/base-files/files/lib/upgrade/common.sh
+++ b/package/base-files/files/lib/upgrade/common.sh
@@ -95,7 +95,7 @@ get_image() { # <source> [ <command> ]
 	if [ -z "$cmd" ]; then
 		local magic="$(dd if="$from" bs=2 count=1 2>/dev/null | hexdump -n 2 -e '1/1 "%02x"')"
 		case "$magic" in
-			1f8b) cmd="zcat";;
+			1f8b) cmd="busybox zcat";;
 			425a) cmd="bzcat";;
 			*) cmd="cat";;
 		esac
-- 
```

fanck0605 committed:
base-files: upgrade: use zcat command provided by busybox

Calling `switch_to_ramfs()` will not copy the gzip executable
(/bin/gzip) to ramfs, but `/bin/zcat` will call `/bin/gzip` when
package gzip is installed, instead of the busybox-supplied zcat.
This will cause `zcat` to fail to find `gzip`, then cause the
sysupgrade to fail. Adding the `busybox` prefix here will solve
the problem.

Signed-off-by: Chuck Fan <fanck0605@qq.com>

fanck0605 commented:
@zhanhb @adschm updated

aparcar commented:
Merged, thank you!
